### PR TITLE
feat(payment): customer ip field inclusion

### PR DESCRIPTION
**Pull request by prajjwalkumar17**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description
Addition of customer_ip field to be taken from either x-forwarded-for or from request itself.

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Motivation and Context
Will help Sdk team for minimizing latency as currently they are using 3rd party service for requesting the Ip addr.

## How did you test it?
Tested it manually

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/router/src/routes/payments.rs [hunk 1]
```diff
@@ -321,6 +322,14 @@ pub async fn payments_confirm(
         return http_not_implemented();
     };
 
+    if payload.customer_ip.is_none() {
+        payload.customer_ip = req
+            .headers()
+            .get("x-forwarded-for")
```
Comment (Narayanbhat166) [inline crates/router/src/routes/payments.rs:328]: add this string literal to `headers`.
https://github.com/juspay/hyperswitch/blob/6f1a1a3bcb7466738b018a773a5fe4adc5cf2887/crates/router/src/lib.rs#L47

File: crates/router/src/routes/payments.rs [hunk 2]
```diff
@@ -321,6 +322,14 @@ pub async fn payments_confirm(
...
+            .and_then(|val| val.to_str().ok().and_then(|ip| ip.split(',').next()))
```
Comment (Narayanbhat166) [inline crates/router/src/routes/payments.rs:329]: Add the comments as to what this transformation does. Provide the original ip address format and the transformed one after running this logic

**Review by Narayanbhat166 — CHANGES_REQUESTED** [review]

File: crates/api_models/src/payments.rs [hunk 3]
```diff
@@ -114,6 +114,10 @@ pub struct PaymentsRequest {
     #[schema(default = false, example = true)]
     pub confirm: Option<bool>,
 
+    ///To check whether customer_ip_address is available or not
+    #[schema(default = false, example = "192.168.1.6")]
+    pub customer_ip: Option<Secret<String, pii::IpAddress>>,
```
[the remark this answers is not in the document]
Comment (vspecky) [reply crates/api_models/src/payments.rs]: Removed. We will make use of `BrowserInformation.ip_address`

**Review by jarnura — APPROVED** [review]

**Review by vspecky — APPROVED** [review]

File: crates/router/src/connector/adyen/transformers.rs [hunk 4]
```diff
@@ -839,25 +841,73 @@ fn get_recurring_processing_model(
...
-                color_depth: info.color_depth,
-                user_agent: info.user_agent.clone(),
-                time_zone_offset: info.time_zone,
-                java_enabled: info.java_enabled,
+            .map(|info| {
+                Ok(AdyenBrowserInfo {
+                    accept_header: info
+                        .accept_header
+                        .clone()
+                        .get_required_value("accept_header")
```
Comment (Narayanbhat166) [inline crates/router/src/connector/adyen/transformers.rs:858]: you can use this function with `.ok_or_else()`
https://github.com/juspay/hyperswitch/blob/0575b26b4fc229e92aef179146dfd561a9ee7f27/crates/router/src/connector/utils.rs#L24

File: crates/router/src/connector/adyen/transformers.rs [hunk 5]
```diff
@@ -839,25 +841,73 @@ fn get_recurring_processing_model(
...
-                screen_width: info.screen_width,
-                color_depth: info.color_depth,
-                user_agent: info.user_agent.clone(),
-                time_zone_offset: info.time_zone,
-                java_enabled: info.java_enabled,
+            .map(|info| {
+                Ok(AdyenBrowserInfo {
+                    accept_header: info
+                        .accept_header
+                        .clone()
```
Comment (Narayanbhat166) [inline crates/router/src/connector/adyen/transformers.rs:857]: are these clones necessary?

File: crates/router/src/routes/payments.rs [hunk 6]
```diff
@@ -321,6 +323,10 @@ pub async fn payments_confirm(
         return http_not_implemented();
     };
 
+    if let Err(err) = helpers::populate_ip_into_browser_info(&req, &mut payload) {
+        return api::log_and_return_error_response(err);
+    }
```
Comment (Narayanbhat166) [inline crates/router/src/routes/payments.rs:328]: Why should we fail this request if ip address population fails?

File: crates/router/src/connector/zen/transformers.rs [hunk 7]
```diff
@@ -345,26 +345,76 @@ fn get_item_object(
...
-        screen_height: browser_info.screen_height.to_string(),
-        screen_width: browser_info.screen_width.to_string(),
-        timezone: browser_info.time_zone.to_string(),
-        accept_header: browser_info.accept_header.clone(),
+        color_depth: browser_info
+            .color_depth
+            .get_required_value("color_depth")
+            .change_context(errors::ConnectorError::MissingRequiredField {
+                field_name: "color_depth",
+            })?
```
Comment (Narayanbhat166) [inline crates/router/src/connector/zen/transformers.rs:379]: Are all these fields really mandatory? can you check the api documentation and confirm this?

**Review by Narayanbhat166 — CHANGES_REQUESTED** [review]

**Review by Narayanbhat166 — APPROVED** [review]
The comments will be addressed in #1398

**Pull request merged**